### PR TITLE
Allow specialization constant overrides in config

### PR DESCRIPTION
**Pull request by pchome**
Only simple types (bool, int, float) for now.

Example:
```ini
reshade = reshade-shaders/Pixelate.fx

# type  = slider
# min   = 2
# max   = 48
# Cell Size
#cell_size = 4
cell_size = 2

# type  = slider
# min   = 0.0
# max   = 1.0
# Smoothness
#avg_amount = 0.333
avg_amount = 1.0
```

### CODE REVIEW

**Comment by DadSchoorse** [comment]
Interesting. The my question is if this would be easier for the end user than editing the shader file directly. Also different shaders could have a spec constant with the same name and the user probably does not want to set all of them to the same value.

File: src/effect_reshade.cpp [hunk 1]
```diff
@@ -332,14 +332,78 @@ namespace vkBasalt
...
+	    // Configure effect
+            std::vector<VkSpecializationMapEntry> specMapEntrys;
+            std::vector<char> specData;
+
+            uint32_t spec_id = 0;
+            uint32_t offset = 0;
+
+            uint32_t bvalue, uvalue;
+            int32_t ivalue;
+            float fvalue;
```
Comment (DadSchoorse) [inline src/effect_reshade.cpp]: Why aren't these defined in the loop?

**Comment by pchome** [comment]
> The my question is if this would be easier for the end user than editing the shader file directly.
1. Yes, if the shader file is in git managed directory.
2. Yes, if it's me, who don't want to keep and maintain additional patch, and who generating config file automatically with protonfixes (e.g `util.set_vkbasalt_option('Power', '3.5')`) :smile: 

Anyway, they don't forced to configure effects or add every option in other case. It's even better when all unused options will be commented out, then their default values will be used.

**Comment by DadSchoorse** [comment]
could the name in the config file be changed to something like `*effect_name*.*spec_constant_name*`

File: src/effect_reshade.cpp (hunk 1 above)
Comment (pchome) [reply src/effect_reshade.cpp]: Placed where `std::vector<std::variant<float,uint32_t,int32_t>>` was in my earlier attempt.
Better refactor this code late, if you going to accept it. Maybe into separate helper function. Currently it's just a hack.

**Comment by pchome** [comment]
I guess per-effect config file support would be better.
E.g. if I use CRT.fx, then check if there is CRT.conf in current workdir. This will keep main config file cleaner, and will allow to have per-game settings.

But shouldn't be a problem to cut `*effect_name*.` string too, if there's such settings in the main config file.

**Comment by DadSchoorse** [comment]
> I guess per-effect config file support would be better.

Yes I think that would be the best solution .

**Comment by pchome** [comment]
> effect1.cell_size

Also, the problem is that some effects can have long names with spaces. E.g. `PD80 Contrasts & ColorFX.fx`.

**Comment by DadSchoorse** [comment]
maybe we should support quotes?
```
effect1 = "reshade-shaders/PD80 Contrasts & ColorFX.fx"
```

**Comment by pchome** [comment]
Or maybe INI file format. It will allow to use sections, and spaces in section names. Should be possible to do something like this:
```ini
[reshade-shaders/PD80 Contrasts & ColorFX.fx]
option = 1
option2 = 2.2

[reshade-shaders/CRT.fx]
option = 1
```

**Comment by pchome** [comment]
BTW, similar thing (dynamic spec constants) can be done for your "simple effects", what (I guess) will let you merge all separate effects code. The downside is that additional dependency will be needed, e.g. spirv-cross.

Sample code for config file generator:
```c++
#include "spirv_cross/spirv_cross.hpp"
...
  // Read SPIR-V file
  std::ifstream rawCFile(spvFile, std::ios::binary);
  std::vector<char> spirv_data((std::istreambuf_iterator<char>(rawCFile)), std::istreambuf_iterator<char>());

  // evil magic
  spirv_cross::Compiler comp(std::move(*reinterpret_cast<std::vector<uint32_t>*>(&spirv_data)));

  for (auto& con : comp.get_specialization_constants()) {
    spirv_cross::SPIRConstant cnst = comp.get_constant(con.id);
    output
      << "c: " << con.constant_id
      << " : " << comp.get_name(con.id)
      << " = " << cnst.scalar()
      << " t:" << cnst.constant_type
    << std::endl;
  }
```
Since you already have SPIR-V file loaded, you could feed it's data to spirv_cross::Compiler and get what you need from it.

LDFLAGS += `-lspirv-cross-core -lspirv-cross-cpp` required.

But that just FYI.

**Comment by crosire** [comment]
INI format would have the advantage that you could plug in existing ReShade preset files: ReShade generates files of the format:
```
Techniques=Comma,separated,list,of,active,technique,names

[Effect1.fx]
Variable1=Value1
Variable2=Value2

[Effect2.fx]
Variable1=Value1
Variable2=Value2
```
Where the INI categories are effect file names (file name only, no directory, but with fx extension) and the `Techniques` value is e.g. `Techniques=SMAA,AspectRatioPS` (so not file names, but names of any techniques inside the effect files that should be used and rendered).

But that would mean adding an INI parser, which, while not overly complicated, may not be desireable.

**Comment by DadSchoorse** [comment]
I'm merging this for now but the config system needs a overhaul

**Pull request merged**